File: js/src/backends/memory.js
/**
 * In-memory LinkStore implementation.
 *
 * Provides O(1) access for development and testing scenarios.
 * This implementation uses JavaScript Map for efficient lookups by ID
 * and supports link deduplication (identical structures share the same ID).
 *
 * @module backends/memory
 */

/*
 * Note: The require-await warnings are intentional.
 * This implementation uses async methods for API consistency with other backends
 * (e.g., file-based or network-based stores) even though the in-memory operations
 * are synchronous. This allows for seamless backend swapping.
 */
/* eslint-disable require-await */

import { getLinkId, createLink, matchesPattern } from '../index.js';

/**
 * In-memory implementation of the LinkStore interface.
 *
 * Features:
 * - O(1) lookups by ID using Map
 * - Link deduplication (identical source/target pairs share same ID)
 * - Pattern matching with wildcard support via Any symbol
 * - Async API for consistency with other backends
 *
 * @implements {import('../index.d.ts').LinkStore}
 *
 * @example
 * const store = new MemoryLinkStore();
 *
 * // Create a link
 * const link = await store.create('hello', 'world');
 * console.log(link); // { id: 1, source: 'hello', target: 'world' }
 *
 * // Find links
 * const results = await store.find({ source: 'hello', target: Any });
 *
 * // Deduplication - returns existing link
 * const duplicate = await store.create('hello', 'world');
 * console.log(duplicate.id === link.id); // true
 */
export class MemoryLinkStore {
  /**
   * Creates a new MemoryLinkStore instance.
   */
  constructor() {
    /**
     * Primary storage - maps link ID to link object.
     * @type {Map<import('../index.d.ts').LinkId, import('../index.d.ts').Link>}
     * @private
     */
    this._links = new Map();

    /**
     * Index for deduplication - maps "source:target" hash to link ID.
     * @type {Map<string, import('../index.d.ts').LinkId>}
     * @private
     */
    this._deduplicationIndex = new Map();

    /**
     * Counter for generating sequential IDs.
     * @type {number}
     * @private
     */
    this._nextId = 1;
  }

  /**
   * Generates a unique hash key for deduplication based on source and target.
   * @param {import('../index.d.ts').LinkRef} source - The source reference
   * @param {import('../index.d.ts').LinkRef} target - The target reference
   * @returns {string} A hash key for the source/target pair
   * @private
   */
  _getDeduplicationKey(source, target) {
    const sourceId = getLinkId(source);
    const targetId = getLinkId(target);
    // Convert to string representation with type prefix to handle different types
    // and prevent collisions (e.g., number 1 vs string "1" vs bigint 1n)
    const serializeId = (id) => {
      if (typeof id === 'bigint') {
        return `bigint:${id.toString()}`;
      }
      if (typeof id === 'number') {
        return `number:${id}`;
      }
      return `string:${id}`;
    };
    return `${serializeId(sourceId)}|${serializeId(targetId)}`;
  }

  /**
   * Creates a new link from source to target.
   *
   * If a link with identical source and target already exists,
   * returns the existing link (deduplication).
   *
   * @param {import('../index.d.ts').LinkRef} source - The source reference
   * @param {import('../index.d.ts').LinkRef} target - The target reference
   * @returns {Promise<import('../index.d.ts').Link>} The created or existing link
   *
   * @example
   * const link = await store.create(1, 2);
   * // { id: 1, source: 1, target: 2 }
   *
   * const duplicate = await store.create(1, 2);
   * // Returns the same link (deduplication)
   * console.log(duplicate.id === link.id); // true
   */
  async create(source, target) {
    // Check for existing link with same source/target (deduplication)
    const dedupeKey = this._getDeduplicationKey(source, target);
    const existingId = this._deduplicationIndex.get(dedupeKey);

    if (existingId !== undefined) {
      return this._links.get(existingId);
    }

    // Create new link
    const id = this._nextId++;
    const link = createLink(id, source, target);

    this._links.set(id, link);
    this._deduplicationIndex.set(dedupeKey, id);

    return link;
  }

  /**
   * Creates a new link with additional values (universal link).
   *
   * Note: Links with values are NOT deduplicated since values are
   * considered part of the link's identity.
   *
   * @param {import('../index.d.ts').LinkRef} source - The source reference
   * @param {import('../index.d.ts').LinkRef} target - The target reference
   * @param {readonly import('../index.d.ts').LinkRef[]} values - Additional values
   * @returns {Promise<import('../index.d.ts').Link>} The created link
   */
  async createWithValues(source, target, values) {
    // Universal links with values are not deduplicated
    const id = this._nextId++;
    const link = createLink(id, source, target, [...values]);

    this._links.set(id, link);

    return link;
  }

  /**
   * Retrieves a link by its ID.
   *
   * @param {import('../index.d.ts').LinkId} id - The ID of the link to retrieve
   * @returns {Promise<import('../index.d.ts').Link | null>} The link if found, null otherwise
   *
   * @example
   * const link = await store.get(42);
   * if (link) {
   *   console.log(`Found: ${link.source} -> ${link.target}`);
   * }
   */
  async get(id) {
    return this._links.get(id) ?? null;
  }

  /**
   * Checks if a link with the given ID exists.
   *
   * @param {import('../index.d.ts').LinkId} id - The ID to check
   * @returns {Promise<boolean>} True if the link exists
   *
   * @example
   * if (await store.exists(42)) {
   *   console.log('Link 42 exists');
   * }
   */
  async exists(id) {
    return this._links.has(id);
  }

  /**
   * Finds all links matching the given pattern.
   *
   * @param {import('../index.d.ts').LinkPattern} pattern - The pattern to match
   * @returns {Promise<import('../index.d.ts').Link[]>} Array of matching links
   *
   * @example
   * // Find all links with source = 5
   * const links = await store.find({ source: 5 });
   *
   * // Find all links with any source, target = 10
   * const links2 = await store.find({ source: Any, target: 10 });
   */
  async find(pattern) {
    const results = [];

    for (const link of this._links.values()) {
      if (matchesPattern(link, pattern)) {
        results.push(link);
      }
    }

    return results;
  }

  /**
   * Counts links matching the given pattern.
   *
   * @param {import('../index.d.ts').LinkPattern} [pattern] - Optional pattern to match
   * @returns {Promise<number>} Count of matching links
   *
   * @example
   * const total = await store.count();
   * const withSource5 = await store.count({ source: 5 });
   */
  async count(pattern) {
    if (!pattern) {
      return this._links.size;
    }

    let count = 0;
    for (const link of this._links.values()) {
      if (matchesPattern(link, pattern)) {
        count++;
      }
    }

    return count;
  }

  /**
   * Updates an existing link's source and/or target.
   *
   * @param {import('../index.d.ts').LinkId} id - The ID of the link to update
   * @param {import('../index.d.ts').LinkRef} source - New source reference
   * @param {import('../index.d.ts').LinkRef} target - New target reference
   * @returns {Promise<import('../index.d.ts').Link>} The updated link
   * @throws {Error} If the link with given ID does not exist
   *
   * @example
   * const updated = await store.update(1, 10, 20);
   * console.log(updated); // { id: 1, source: 10, target: 20 }
   */
  async update(id, source, target) {
    const existingLink = this._links.get(id);

    if (!existingLink) {
      throw new Error(`Link with id ${id} not found`);
    }

    // Remove old deduplication entry
    const oldDedupeKey = this._getDeduplicationKey(
      existingLink.source,
      existingLink.target
    );
    this._deduplicationIndex.delete(oldDedupeKey);

    // Create updated link (preserving values if any)
    const updatedLink = existingLink.values
      ? createLink(id, source, target, [...existingLink.values])
      : createLink(id, source, target);

    this._links.set(id, updatedLink);

    // Add new deduplication entry (only for links without values)
    if (!existingLink.values) {
      const newDedupeKey = this._getDeduplicationKey(source, target);
      this._deduplicationIndex.set(newDedupeKey, id);
    }

    return updatedLink;
  }

  /**
   * Deletes a link by its ID.
   *
   * @param {import('../index.d.ts').LinkId} id - The ID of the link to delete
   * @returns {Promise<boolean>} True if deleted, false if not found
   *
   * @example
   * const deleted = await store.delete(42);
   * if (deleted) {
   *   console.log('Link 42 was deleted');
   * }
   */
  async delete(id) {
    const link = this._links.get(id);

    if (!link) {
      return false;
    }

    // Remove from deduplication index (only for links without values)
    if (!link.values) {
      const dedupeKey = this._getDeduplicationKey(link.source, link.target);
      this._deduplicationIndex.delete(dedupeKey);
    }

    this._links.delete(id);
    return true;
  }

  /**
   * Deletes all links matching the given pattern.
   *
   * @param {import('../index.d.ts').LinkPattern} pattern - The pattern to match
   * @returns {Promise<number>} Count of deleted links
   *
   * @example
   * const count = await store.deleteMatching({ source: 5 });
   * console.log(`Deleted ${count} links`);
   */
  async deleteMatching(pattern) {
    const toDelete = [];

    for (const [id, link] of this._links.entries()) {
      if (matchesPattern(link, pattern)) {
        toDelete.push(id);
      }
    }

    for (const id of toDelete) {
      await this.delete(id);
    }

    return toDelete.length;
  }

  /**
   * Iterates over all links matching the given pattern.
   *
   * @param {import('../index.d.ts').LinkPattern} [pattern] - Optional pattern to filter
   * @returns {AsyncIterable<import('../index.d.ts').Link>} Async iterable of matching links
   *
   * @example
   * for await (const link of store.iterate({ source: 5 })) {
   *   console.log(link);
   * }
   */
  async *iterate(pattern) {
    for (const link of this._links.values()) {
      if (!pattern || matchesPattern(link, pattern)) {
        yield link;
      }
    }
  }

  /**
   * Removes all links from the store.
   *
   * @returns {Promise<void>}
   *
   * @example
   * await store.clear();
   * console.log(await store.count()); // 0
   */
  async clear() {
    this._links.clear();
    this._deduplicationIndex.clear();
    this._nextId = 1;
  }
}

File: rust/src/backends/memory.rs
//! In-memory storage backend for links-queue.
//!
//! This module provides [`MemoryLinkStore`], an in-memory implementation of the
//! [`LinkStore`] trait. It offers O(1) lookups by ID and supports link deduplication,
//! pattern matching, and thread-safe operations.
//!
//! # Features
//!
//! - **O(1) lookups**: Uses `HashMap` for constant-time access by ID.
//! - **Deduplication**: Identical link structures (same source/target) share the same ID.
//! - **Pattern matching**: Supports wildcard queries via [`LinkPattern`].
//! - **Thread-safe**: Implements `Send + Sync` for concurrent access.
//!
//! # Design
//!
//! This implementation follows patterns from [doublets-rs](https://github.com/linksplatform/doublets-rs)
//! while adapting to Rust idioms and the links-queue API contract.
//!
//! # Example
//!
//! ```rust
//! use links_queue::{MemoryLinkStore, LinkStore, LinkRef, LinkPattern, Any};
//!
//! let mut store = MemoryLinkStore::<u64>::new();
//!
//! // Create a link
//! let link = store.create(LinkRef::Id(2), LinkRef::Id(3)).unwrap();
//! assert_eq!(link.id, 1); // Auto-generated ID
//!
//! // Deduplication: same source/target returns existing link
//! let duplicate = store.create(LinkRef::Id(2), LinkRef::Id(3)).unwrap();
//! assert_eq!(duplicate.id, link.id);
//!
//! // Find by pattern
//! let results = store.find(&LinkPattern::with_source(LinkRef::Id(2)));
//! assert_eq!(results.len(), 1);
//!
//! // Update a link
//! let updated = store.update(link.id, LinkRef::Id(4), LinkRef::Id(5)).unwrap();
//! assert_eq!(updated.source_id(), 4);
//!
//! // Delete a link
//! assert!(store.delete(link.id));
//! assert!(!store.exists(link.id));
//! ```

use std::collections::HashMap;
use std::hash::{Hash, Hasher};

use crate::{Link, LinkError, LinkPattern, LinkRef, LinkResult, LinkStore, LinkType};

// =============================================================================
// Helper Types for Deduplication
// =============================================================================

/// A key for deduplicating links based on their structure (source/target/values).
///
/// This is used internally to detect when a link with the same structure
/// already exists, allowing us to return the existing ID instead of creating
/// a duplicate.
#[derive(Debug, Clone)]
struct LinkKey<T: LinkType> {
    source_id: T,
    target_id: T,
    values: Option<Vec<T>>,
}

impl<T: LinkType> LinkKey<T> {
    fn from_refs(
        source: &LinkRef<T>,
        target: &LinkRef<T>,
        values: Option<&Vec<LinkRef<T>>>,
    ) -> Self {
        Self {
            source_id: source.get_id(),
            target_id: target.get_id(),
            values: values.map(|v| v.iter().map(LinkRef::get_id).collect()),
        }
    }
}

impl<T: LinkType> PartialEq for LinkKey<T> {
    fn eq(&self, other: &Self) -> bool {
        self.source_id == other.source_id
            && self.target_id == other.target_id
            && self.values == other.values
    }
}

impl<T: LinkType> Eq for LinkKey<T> {}

impl<T: LinkType> Hash for LinkKey<T> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.source_id.hash(state);
        self.target_id.hash(state);
        self.values.hash(state);
    }
}

// =============================================================================
// MemoryLinkStore Implementation
// =============================================================================

/// In-memory storage backend implementing the [`LinkStore`] trait.
///
/// Provides O(1) access by ID using a `HashMap`, automatic ID generation,
/// and link deduplication (identical structures share the same ID).
///
/// # Type Parameters
///
/// * `T` - The link ID type (must implement [`LinkType`])
///
/// # Thread Safety
///
/// `MemoryLinkStore` is `Send + Sync`, making it safe to share between threads.
/// However, it does not provide internal synchronization - callers must handle
/// concurrent access (e.g., using `Mutex` or `RwLock`) if needed.
///
/// # Examples
///
/// ## Basic Usage
///
/// ```rust
/// use links_queue::{MemoryLinkStore, LinkStore, LinkRef};
///
/// let mut store = MemoryLinkStore::<u64>::new();
///
/// // Create links
/// let link1 = store.create(LinkRef::Id(10), LinkRef::Id(20)).unwrap();
/// let link2 = store.create(LinkRef::Id(20), LinkRef::Id(30)).unwrap();
///
/// // Retrieve by ID
/// let retrieved = store.get(link1.id).unwrap();
/// assert_eq!(retrieved.source_id(), 10);
///
/// // Count total links
/// assert_eq!(store.total_count(), 2);
/// ```
///
/// ## Deduplication
///
/// ```rust
/// use links_queue::{MemoryLinkStore, LinkStore, LinkRef};
///
/// let mut store = MemoryLinkStore::<u64>::new();
///
/// let link1 = store.create(LinkRef::Id(1), LinkRef::Id(2)).unwrap();
/// let link2 = store.create(LinkRef::Id(1), LinkRef::Id(2)).unwrap();
///
/// // Same structure returns same ID
/// assert_eq!(link1.id, link2.id);
/// assert_eq!(store.total_count(), 1);
/// ```
#[derive(Debug)]
pub struct MemoryLinkStore<T: LinkType> {
    /// Primary storage: maps link IDs to link data.
    links: HashMap<T, Link<T>>,

    /// Reverse index: maps link structure (source/target) to ID for deduplication.
    dedup_index: HashMap<LinkKey<T>, T>,

    /// Counter for generating unique IDs.
    next_id: T,
}

impl<T: LinkType> Default for MemoryLinkStore<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T: LinkType> MemoryLinkStore<T> {
    /// Creates a new, empty in-memory link store.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use links_queue::{MemoryLinkStore, LinkStore};
    ///
    /// let store = MemoryLinkStore::<u64>::new();
    /// assert_eq!(store.total_count(), 0);
    /// ```
    #[must_use]
    pub fn new() -> Self {
        Self {
            links: HashMap::new(),
            dedup_index: HashMap::new(),
            next_id: T::zero(),
        }
    }

    /// Creates a new link store with pre-allocated capacity.
    ///
    /// Use this when you know approximately how many links will be stored
    /// to avoid reallocations.
    ///
    /// # Arguments
    ///
    /// * `capacity` - The number of links to pre-allocate space for
    ///
    /// # Examples
    ///
    /// ```rust
    /// use links_queue::MemoryLinkStore;
    ///
    /// let store = MemoryLinkStore::<u64>::with_capacity(1000);
    /// ```
    #[must_use]
    pub fn with_capacity(capacity: usize) -> Self {
        Self {
            links: HashMap::with_capacity(capacity),
            dedup_index: HashMap::with_capacity(capacity),
            next_id: T::zero(),
        }
    }

    /// Removes all links from the store.
    ///
    /// This clears both the primary storage and the deduplication index,
    /// but does NOT reset the ID counter. New links will continue with
    /// the next available ID.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use links_queue::{MemoryLinkStore, LinkStore, LinkRef};
    ///
    /// let mut store = MemoryLinkStore::<u64>::new();
    /// store.create(LinkRef::Id(1), LinkRef::Id(2)).unwrap();
    /// store.create(LinkRef::Id(3), LinkRef::Id(4)).unwrap();
    /// assert_eq!(store.total_count(), 2);
    ///
    /// store.clear();
    /// assert_eq!(store.total_count(), 0);
    ///
    /// // ID counter continues from where it left off
    /// let new_link = store.create(LinkRef::Id(5), LinkRef::Id(6)).unwrap();
    /// assert_eq!(new_link.id, 3); // Not 1
    /// ```
    pub fn clear(&mut self) {
        self.links.clear();
        self.dedup_index.clear();
    }

    /// Resets the store to its initial state, including the ID counter.
    ///
    /// Unlike [`clear()`], this also resets the ID counter to zero.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use links_queue::{MemoryLinkStore, LinkStore, LinkRef};
    ///
    /// let mut store = MemoryLinkStore::<u64>::new();
    /// store.create(LinkRef::Id(1), LinkRef::Id(2)).unwrap();
    ///
    /// store.reset();
    ///
    /// let new_link = store.create(LinkRef::Id(5), LinkRef::Id(6)).unwrap();
    /// assert_eq!(new_link.id, 1); // Starts from 1 again
    /// ```
    pub fn reset(&mut self) {
        self.links.clear();
        self.dedup_index.clear();
        self.next_id = T::zero();
    }

    /// Returns the capacity of the underlying storage.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use links_queue::MemoryLinkStore;
    ///
    /// let store = MemoryLinkStore::<u64>::with_capacity(100);
    /// assert!(store.capacity() >= 100);
    /// ```
    #[must_use]
    pub fn capacity(&self) -> usize {
        self.links.capacity()
    }

    /// Generates the next unique ID.
    ///
    /// IDs start at 1 (not 0) to allow 0 to represent "nothing" or "null".
    fn generate_id(&mut self) -> T {
        // Increment first to start from 1, not 0
        self.next_id = self.next_id.increment();
        self.next_id
    }

    /// Creates a link and adds it to storage.
    fn create_link_internal(
        &mut self,
        source: LinkRef<T>,
        target: LinkRef<T>,
        values: Option<Vec<LinkRef<T>>>,
    ) -> Link<T> {
        let key = LinkKey::from_refs(&source, &target, values.as_ref());

        // Check for existing link with same structure (deduplication)
        if let Some(&existing_id) = self.dedup_index.get(&key) {
            if let Some(existing_link) = self.links.get(&existing_id) {
                return existing_link.clone();
            }
        }

        // Generate new ID and create link
        let id = self.generate_id();
        let link = if let Some(vals) = values {
            Link::with_values(id, source, target, vals)
        } else {
            Link::new(id, source, target)
        };

        // Store link and update dedup index
        self.links.insert(id, link.clone());
        self.dedup_index.insert(key, id);

        link
    }
}

// =============================================================================
// LinkStore Trait Implementation
// =============================================================================

impl<T: LinkType> LinkStore<T> for MemoryLinkStore<T> {
    fn create(&mut self, source: LinkRef<T>, target: LinkRef<T>) -> LinkResult<T, Link<T>> {
        Ok(self.create_link_internal(source, target, None))
    }

    fn create_with_values(
        &mut self,
        source: LinkRef<T>,
        target: LinkRef<T>,
        values: Vec<LinkRef<T>>,
    ) -> LinkResult<T, Link<T>> {
        let values_opt = if values.is_empty() {
            None
        } else {
            Some(values)
        };
        Ok(self.create_link_internal(source, target, values_opt))
    }

    fn get(&self, id: T) -> Option<&Link<T>> {
        self.links.get(&id)
    }

    fn find(&self, pattern: &LinkPattern<T>) -> Vec<&Link<T>> {
        self.links
            .values()
            .filter(|link| pattern.matches(link))
            .collect()
    }

    fn update(&mut self, id: T, source: LinkRef<T>, target: LinkRef<T>) -> LinkResult<T, Link<T>> {
        // Check if link exists
        let old_link = self.links.get(&id).ok_or(LinkError::NotFound(id))?.clone();

        // Remove old dedup entry
        let old_key =
            LinkKey::from_refs(&old_link.source, &old_link.target, old_link.values.as_ref());
        self.dedup_index.remove(&old_key);

        // Check if new structure already exists (would create duplicate)
        let new_key = LinkKey::from_refs(&source, &target, old_link.values.as_ref());
        if let Some(&existing_id) = self.dedup_index.get(&new_key) {
            if existing_id != id {
                // Restore old dedup entry before returning error
                self.dedup_index.insert(old_key, id);
                return Err(LinkError::AlreadyExists(existing_id));
            }
        }

        // Update the link
        let updated_link = if let Some(vals) = old_link.values {
            Link::with_values(id, source, target, vals)
        } else {
            Link::new(id, source, target)
        };

        self.links.insert(id, updated_link.clone());
        self.dedup_index.insert(new_key, id);

        Ok(updated_link)
    }

    fn delete(&mut self, id: T) -> bool {
        if let Some(link) = self.links.remove(&id) {
            let key = LinkKey::from_refs(&link.source, &link.target, link.values.as_ref());
            self.dedup_index.remove(&key);
            true
        } else {
            false
        }
    }

    fn delete_matching(&mut self, pattern: &LinkPattern<T>) -> usize {
        // Collect IDs to delete (to avoid borrowing issues)
        let ids_to_delete: Vec<T> = self
            .links
            .iter()
            .filter(|(_, link)| pattern.matches(link))
            .map(|(&id, _)| id)
            .collect();

        let count = ids_to_delete.len();
        for id in ids_to_delete {
            self.delete(id);
        }
        count
    }

    fn iter<'a>(
        &'a self,
        pattern: &'a LinkPattern<T>,
    ) -> Box<dyn Iterator<Item = &'a Link<T>> + 'a> {
        Box::new(
            self.links
                .values()
                .filter(move |link| pattern.matches(link)),
        )
    }

    fn iter_all(&self) -> Box<dyn Iterator<Item = &Link<T>> + '_> {
        Box::new(self.links.values())
    }
}
